apps/query-ui/queryui/util.py

```diff
@@ -242,30 +243,73 @@ def readdata(self):
                     data[col].append(row.get(col))
             self.df = pd.DataFrame(data)
 
-    def show(self):
+    def show(self, node_descriptions: dict[str, str]):
         """Render the trace data."""
+        st.subheader(f"Node {self.node_id}")
+        st.markdown(f"*Description: {node_descriptions.get(self.node_id) or 'n/a'}*")
         if self.df is None or not len(self.df):
-            st.write(f"Result of node {self.node_id} — :red[0] documents")
+            st.write(":red[0] documents")
             st.write("No data.")
             return
 
         all_columns = list(self.df.columns)
         column_order = [c for c in self.COLUMNS if c in all_columns]
         column_order += [c for c in all_columns if c not in column_order]
-        st.write(f"Result of node {self.node_id} — **{len(self.df)}** documents")
+        st.write(f"**{len(self.df)}** documents")
         st.dataframe(self.df, column_order=column_order)
 
 
+class QueryMetadataTrace:
+    """Helper class to read and display metadata about a query."""
+
+    def __init__(self, metadata_dir: str):
+        self.metadata_dir = metadata_dir
+        self.query_plan = None
+        self.readdata()
+
+    def readdata(self):
+        f = os.path.join(self.metadata_dir, "query_plan.pickle")
+        if os.path.isfile(f):
+            with open(f, "rb") as file:
+                self.query_plan = pickle.load(file)
+
+    def get_node_to_description(self) -> dict[str, str]:
+        if self.query_plan is None:
+            return {}
+        result = dict()
+        for node_id, node in self.query_plan.nodes.items():
+            result[str(node_id)] = node.description
+        return result
+
+    def show(self):
+        if self.query_plan is not None:
+            st.write(f"Query: {self.query_plan.query}")
+            st.write(self.query_plan)
+        else:
+            st.write("No query plan found")
+
+
 class QueryTrace:
     """Helper class used to read and display query traces."""
 
     def __init__(self, trace_dir: str):
         self.trace_dir = trace_dir
-        self.node_traces = [QueryNodeTrace(trace_dir, node_id) for node_id in sorted(os.listdir(self.trace_dir))]
+        self.node_traces = []
+        for dir in sorted(os.listdir(self.trace_dir)):
+            if "metadata" not in dir:
+                self.node_traces += [QueryNodeTrace(trace_dir, dir)]
+            self.metadata = QueryMetadataTrace(self.trace_dir + "/" + dir)
```

Comment:
I don't think we need this extra class and indirection here. It's not a 'trace', and 'metadata' as a term is so heavily overused in our entire codebase that I would prefer to avoid it if possible :-)

Could you just have this function look for and read the file `query_plan.json` and set `self.query_plan` to the resulting object? We also don't need the `node_descriptions` map, I think, since it's easy enough to pluck the node we care about out of the `self.query_plan.nodes` map.

Reply:
Done